### PR TITLE
Allow different degrees for different parameters in Fourier tools

### PR DESCRIPTION
**Pull request by dwierichs**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
The `qml.fourier.coefficients` function currently only allows to pass a single degree, giving the maximal frequency considered
in the Fourier transform. This PR changes this to allow a sequence of degrees, one per parameter of the analyzed function.
The functions in `qml.fourier.visualize` are adapted to accept and understand the resulting newly-shaped coefficient arrays.

**Description of the Change:**
Allows the `degree` argument of `coefficients` to be a sequence of integers and computes the Fourier transform with largest frequency in the `i`th parameter given by the `i`th entry of `degree`.
Allows the `coeffs` argument of the visualization tools in `qml.fourier.visualize` to accept coefficient arrays that result from
`coefficients` when called with inhomogeneous `degree`.
Also simplifies a few steps in `coefficients`/`visualize` methods and removes an unnecessary helper function in `fourier.utils`.

**Benefits:**
More flexible Fourier transform. In particular, allows users to save computational effort whenever different input parameters of
the analyzed function enter the function with differing (known) degrees in the Fourier series.

**Possible Drawbacks:**
None. There is no breaking change. Code complexity only slightly increased and decreased a bit in other places.

**Related GitHub Issues:**
N/A

### CODE REVIEW

**Comment by dwierichs** [comment]
@Core in an attempt to tag this PR for review by a core team member :)

File: doc/releases/changelog-dev.md
Comment (albi3ro) [inline doc/releases/changelog-dev.md:7]: I would categorize this as an "Improvement" rather than a "New feature".

**Comment by dwierichs** [comment]
This PR would be good for a review :)
The failing Codecov check should be due to more lines being removed than added here, and hence the reduced coverage rate.

File: pennylane/fourier/visualize.py
Comment (timmysilv) [inline pennylane/fourier/visualize.py:98]:
```suggestion
    # extract flattened data by real part and imaginary part, and respecting negative indices
```

File: pennylane/fourier/coefficients.py
Comment (timmysilv) [inline pennylane/fourier/coefficients.py:252]: can you add a test for this? I'd rather this weren't the case at all, but if you feel it's necessary I wouldn't mind a nice example to demonstrate why

File: doc/releases/changelog-dev.md
Comment (timmysilv) [inline doc/releases/changelog-dev.md:119]: should probably move this PR link to the end of all change text

**Review by timmysilv — COMMENTED** [review]
bringing this back. just some small things, but will be nice to have in soon!

File: pennylane/fourier/coefficients.py
Comment (dwierichs) [reply pennylane/fourier/coefficients.py]: This is being tested already, and the reason for the if-else clause:
When `use_broadcasting=True`, `nvec` is a `tuple` like `(-2, -1, [-3, -2, -1, 0, 1, 2, 3])`, which is a valid numpy index.
The `else` clause (which is the previous behaviour) deals with `nvec` by casting it to an array and multiplying with the `ndarray` `spacing`. The above `tuple` would yield a ragged array, requiring us to do the multiplication via manual iteration over the first axis of `nvec` and `spacing`.

Comment (dwierichs) [reply pennylane/fourier/coefficients.py]: I rewrote the comment to reflect what is actually happening here :)

File: doc/releases/changelog-dev.md
Comment (timmysilv) [reply doc/releases/changelog-dev.md:11]: hmm, I guess most change descriptions I see are either very short or very long so I never really considered what you've got here. I'm convinced! Let's leave it 👍

File: tests/fourier/test_coefficients.py
Comment (mudit2812) [inline tests/fourier/test_coefficients.py:94]: Not blocking but I think a small comment saying something like `# Testing with sequence of degrees` might make the test a bit clearer, because the two added lines are identical apart from the tuple being used for `degree`.

**Review by mudit2812 — COMMENTED** [review]
Thanks for the changes David! The changes look good, and the testing is very thorough. I left a few minor non-blocking suggestions. I'm admittedly still quite unfamiliar with the `qml.fourier` module, so I will refrain from approving for now and review some material before I add an approval.

File: tests/fourier/test_coefficients.py
Comment (dwierichs) [inline tests/fourier/test_coefficients.py:93]:
```suggestion
        assert np.allclose(coeffs, expected_coeffs)
        # Testing with a single-entry sequence of degrees
```

File: tests/fourier/test_coefficients.py
Comment (dwierichs) [inline tests/fourier/test_coefficients.py:90]:
```suggestion
        partial_func = partial(fourier_function, freq_dict)
        # Testing with a single degree provided as integer
```

File: pennylane/fourier/visualize.py
Comment (timmysilv) [inline pennylane/fourier/visualize.py:101]: I might be wrong, but I think you can just do `coeffs.real[nvecs]` and `coeffs.imag[nvecs]` here

**Review by timmysilv — APPROVED** [review]
aesthetic changes only at this point, take em or leave em. LGTM!!

File: doc/releases/changelog-0.26.0.md
Comment (trbromley) [inline doc/releases/changelog-0.26.0.md:513]: Shouldn't it be 3005?

File: pennylane/fourier/coefficients.py
Comment (dwierichs) [reply pennylane/fourier/coefficients.py:242]: Oh gosh :D

**Review by mudit2812 — APPROVED** [review]
Thanks David! Looks good to me :)

File: pennylane/fourier/visualize.py
Comment (dwierichs) [reply pennylane/fourier/visualize.py:101]: Unfortunately this does not work because of the leading axis that is not to be indexed as well.
I tried to make it work by moving the axis before and after slicing but didn't get it to work right away.
Given that this is not a highly used part of the codebase I'd leave it for now :)

**Pull request merged**
